node/events-to-ga/src/utils.js

```diff
@@ -0,0 +1,61 @@
+/**
+ * Throws an error if any of the keys are missing from the object
+ * @param {*} obj
+ * @param {string[]} keys
+ * @throws {Error}
+ */
+export function throwIfMissing(obj, keys) {
+  const missing = [];
+  for (let key of keys) {
+    if (!(key in obj) || !obj[key]) {
+      missing.push(key);
+    }
+  }
+  if (missing.length > 0) {
+    throw new Error(`Missing required fields: ${missing.join(', ')}`);
+  }
+}
+
+/**
+ * Returns whether the request headers and appropriate message are ok
+ * @param {*} req
+ * @returns {}
+ */
+export function verifyHeaders(req) {
+  if (req.headers['x-appwrite-user-id'] == '') {
+    throw new Error(`x-appwrite-user-id value in req.headers is not there`);
+  }
+  if (req.headers['x-appwrite-trigger'] != 'event') {
+    throw new Error(
+      `Not triggered by event but by ${req.headers['x-appwrite-trigger']}`
+    );
+  }
+  if (req.headers['x-appwrite-event'] == '') {
+    throw new Error(`x-appwrite-event value in req.headers is not there`);
+  }
+}
+
+/**
+ * Returns object containing a string denoting the formatted event name to Google Analytics and another object containing params
+ * @param {string} str
+ * @returns {{event_name:string,wildCardObject:{[key:string]:string}}}
+ */
+export function formatIntoGoogleAnalyticsEvent(str) {
+  const oddElemArray = [];
+  const wildCardArray = [];
+  const splitArray = str.split('.');
+
+  for (let i = 0; i < splitArray.length; i++) {
+    if (i % 2 == 0) oddElemArray.push(splitArray[i]);
+    else
+      wildCardArray.push([
+        `${oddElemArray[oddElemArray.length - 1]}Id`,
```

Comment:
Hmm this will result in something like `databasesId`, right? It would be better if it was singular to match how we typically have it.

Reply:
Also, this algorithm won't work for events like `teams.*.memberships.*.update.status`

Reply:
is it because of no 's' in update because that I can resolve but I cant find out a way to get [status,prefs,name,email,password] unless I consider them that these will not go to the wildcard array

Reply:
can I do that

Reply:
Can I assume the length of the words in wildCard array is more than 12 or somethiing to distinguish it

Reply:
> is it because of no 's' in update because that I can resolve but I cant find out a way to get [status,prefs,name,email,password] unless I consider them that these will not go to the wildcard array

Because not every odd element is an ID. Although...I guess it's fine that `status` ends up in the params...

Reply:
> Can I assume the length of the words in wildCard array is more than 12 or somethiing to distinguish it

No, an ID isn't necessarily more than 12 chars.